aten/src/ATen/native/cuda/CUDALoops.cuh

```diff
@@ -144,95 +146,98 @@ namespace modern {
 
 namespace detail {
 
-template <typename func_t, typename array_t, std::size_t... I>
-__device__ inline constexpr decltype(auto) invoke_with_array_impl(func_t f, array_t t, std::index_sequence<I...>)
-{
-    return f(t[I]...);
-}
-template <typename func_t, typename array_t>
-__device__ inline constexpr decltype(auto) invoke_with_array(func_t f, array_t a) {
-  constexpr auto arity = function_traits<func_t>::arity;
-  return invoke_with_array_impl(f, a, std::make_index_sequence<arity>{});
-}
+// The `pointers` converts std::tuple<T1, T2, ....> to std::tuple<T1*, T2*, ....>
 
-namespace arg_type {
+template <typename T>
+struct pointers_helper {};
 
-// We need a way to compute the argument type of a function. But
-// for nullary function, it does not really have an argument type
-// in this case, we still need to return a valid type, but we don't
-// really care what type this is.
+template <typename... types>
+struct pointers_helper<std::tuple<types...>> {
+  using type = std::tuple<types *...>;
+};
 
-struct dont_care {};
+template <typename T>
+using pointers = typename pointers_helper<T>::type;
 
-template <typename func_t, std::size_t arity>
-struct arg_type_helper {
-  using type = typename function_traits<func_t>::template arg<0>::type;
+template<template<int i> typename func, int end, int current=0>
+struct static_unroll {
+  template<typename... Args>
+  static inline C10_HOST_DEVICE void with_args(Args&&... args) {
+    func<current>::apply(std::forward<Args>(args)...);
+    static_unroll<func, end, current+1>::with_args(args...);
+  }
 };
 
-template <typename func_t>
-struct arg_type_helper<func_t, 0> {
-  using type = dont_care;
+template<template<int i> typename func, int end>
+struct static_unroll<func, end, end> {
+  template<typename... Args>
+  static inline C10_HOST_DEVICE void with_args(Args... args) {}
 };
 
-template <typename func_t>
-using type = typename arg_type_helper<func_t, function_traits<func_t>::arity>::type;
-
-}  // namespace arg_type
+template<int i>
+struct can_vectorize_up_to_helper {
+  template <typename array_t, typename traits>
+  static C10_HOST_DEVICE void apply(int &result, array_t pointers, traits _) {
+    using arg_t = std::tuple_element_t<i, typename traits::ArgsTuple>;
```

Comment:
`using arg_t = typename traits::template arg<0>::type`, like it's done in `*_function_traits`? I think it's slightly more readable, but up to you.

Reply:
Yes, this makes more sense.